### PR TITLE
Closes #4953: Adds try again hook for downloads

### LINKED ISSUE #4953: Add try again functionality to downloadFeature
Since we need to access this functionality from other places in Fenix (not just inside the notification service), we need a hook to call `tryAgain`.

### PR DESCRIPTION
**Pull request by sblatz**
---

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [ ] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/feature/downloads/src/main/java/mozilla/components/feature/downloads/manager/AndroidDownloadManager.kt [hunk 1]
```diff
@@ -99,19 +111,15 @@ class AndroidDownloadManager(
      * broadcast receiver if there are no more queued downloads.
      */
     override fun onReceive(context: Context, intent: Intent) {
-        if (queuedDownloads.isEmpty()) unregisterListeners()
```
Comment (sblatz) [inline components/feature/downloads/src/main/java/mozilla/components/feature/downloads/manager/AndroidDownloadManager.kt]: We must remove these so future `onDownloadCompleted` calls get triggered (even if the user just does a `tryAgain`. This actually fixes an existing bug: Download fails, press "try again" in notification, `onDownloadCompleted` does not get called. 

The listeners are still unregistered in `onStop` so there's no leak here.

**Comment by sblatz** [comment]
Added tests, pls approve of me @csadilek 😭

**Comment by csadilek** [comment]
@sblatz looks like the change causes the test `calling registerListener with valid downloadID must call listener after download` to fail now.

**Comment by sblatz** [comment]
bors r=@csadilek

**Comment by sblatz** [comment]
bors retry

**Comment by sblatz** [comment]
bors retry

**Pull request merged**